app/src/rx_fastmetadata_app.c

```diff
@@ -4,12 +4,152 @@
+  }
+
+  uint8_t* m = mmap(NULL, i.st_size, PROT_READ, MAP_SHARED, fd, 0);
+  if (MAP_FAILED == m) {
+    err("%s(%d), mmap '%s' fail.\n", __func__, idx, session->st41_ref_url);
+    close(fd);
+    return -EIO;
+  }
+
+  session->st41_ref_begin = m;
+  session->st41_ref_cursor = m;
+  session->st41_ref_end = m + i.st_size;
+  session->st41_ref_fd = fd;
+
+  info("%s, opening file '%s' success.\n", __func__, session->st41_ref_url);
+  return 0;
+}
+
+static int app_rx_fmd_compare_with_ref(struct st_app_rx_fmd_session* session, void* frame,
+                                       int frame_size) {
+  int ret = -1;
+  uint32_t last_zeros = 0; /* 4 bytes with 0 */
+  uint32_t st41_ref_remaining_length =
+      session ? session->st41_ref_end - session->st41_ref_cursor : 0;
+
+  if (frame_size <= st41_ref_remaining_length) {
+    ret = memcmp(frame, session->st41_ref_cursor, frame_size);
+    if (ret) {
+      session->errors_count++;
+      err("%s() FAIL: reference file comparison (1).\n", __func__);
+    } else {
+      dbg("%s() PASS: reference file comparison (1).\n", __func__);
+    }
+  } else {
+    if (frame_size - st41_ref_remaining_length > 3) {
+      ret = -1;
+      session->errors_count++;
+      err("%s() FAIL: frame_size > ref_remaining_length by %d.\n", __func__,
+          frame_size - st41_ref_remaining_length);
+    } else {
+      ret = memcmp(frame, session->st41_ref_cursor, st41_ref_remaining_length);
+      if (ret) {
+        session->errors_count++;
+        err("%s() FAIL: reference file comparison (2).\n", __func__);
+      } else {
+        dbg("%s() PASS: reference file comparison (2).\n", __func__);
+
+        /* Verify last 0-3 bytes of frame (filled with zero's) */
+        ret = memcmp(&(((uint8_t*)frame)[st41_ref_remaining_length]), (void*)&last_zeros,
+                     frame_size - st41_ref_remaining_length);
+        if (ret) {
+          session->errors_count++;
+          err("%s() FAIL: reference file comparison (3).\n", __func__);
+        } else {
+          dbg("%s() PASS: reference file comparison (3).\n", __func__);
+        }
+      }
+    }
+  }
+
+#ifdef DEBUG /* print out of frame and reference file if error */
+  if (ret) {
+    err("%s() FRAME START>>", __func__);
+    for (int i = 0; i < frame_size; i++) {
+      err("%c", ((char*)frame)[i]);
+    }
+    err("<<END FRAME.\n\n");
+
+    err("%s() REFERENCE START>>", __func__);
+    for (int i = 0; i < frame_size; i++) {
+      err("%c", ((char*)(session->st41_ref_cursor))[i]);
+    }
+    err("<<END REFERENCE.\n\n");
+  }
+#endif /* DEBUG */
+
+  /* calculate new reference frame */
+  session->st41_ref_cursor += frame_size;
+  if (session->st41_ref_cursor >= session->st41_ref_end)
+    session->st41_ref_cursor = session->st41_ref_begin;
+
+  return ret;
+}
+
+static void app_rx_fmd_handle_rtp(struct st_app_rx_fmd_session* s, void* usrptr) {
+  struct st41_rtp_hdr* hdr = (struct st41_rtp_hdr*)usrptr;
+  uint8_t* payload = (uint8_t*)(&hdr[1]);
```

Comment:
we should't do it this way as this will trigger scans 
i don't suppose we have a better structure to cast the usrptr to include 
{
hdr
void* payload  
}
? 

This is the most elegant solution
if possible ofc